### DISCUSSION TITLE
Highlight trailing whitespaces

### DISCUSSION THREAD

**Discussion by horseinthesky** [lukas-reineke/indent-blankline.nvim#762] [Q&A] [answered]
Hello.
There is a `highlight` option for `whitespace`.
Is there a way to highlight trailing whitespaces only? Would be great to see them in red or something.

**Comment by lukas-reineke** [comment, marked as answer]
There is no option, but you can make a hook for this.

This is a simple example, there might be edge cases update this as you need.

```lua
local utils = require "ibl.utils"
local hooks = require "ibl.hooks"

hooks.register(hooks.type.VIRTUAL_TEXT, function(_, bufnr, row, virt_text)
    -- Get the line for this row
    local line = vim.api.nvim_buf_get_lines(bufnr, row, row + 1, false)[1]

    -- Check if it has trailing whitespace
    if #line == 0 or #line ~= #utils.get_whitespace(line) then
        return virt_text
    end

    -- Iterate over the virtual text
    for i, cell in ipairs(virt_text) do
        -- Iterate over each virtual text cell
        for j, highlight_group in ipairs(cell[2]) do
            -- Overwrite any highlight that is whitespace
            if vim.startswith(highlight_group, "@ibl.whitespace") then
                virt_text[i][2][j] = "Error" -- Change this to whatever highlight group you want
            end
        end
    end
    return virt_text
end)
```

**Reply by horseinthesky** [reply]
Hm. Looks thicky =)

Can't actually understand the details but it doesn't work for me.

Lazy config
```
  {
    "lukas-reineke/indent-blankline.nvim",
    event = "BufRead",
    config = function()
      require("ibl").setup {
        scope = {
          show_start = false,
          show_end = false,
        },
        indent = {
          char = { "", "¦", "┆", "┊" },
        },
      }
      local utils = require "ibl.utils"
      local hooks = require "ibl.hooks"

      hooks.register(hooks.type.VIRTUAL_TEXT, function(_, bufnr, row, virt_text)
        -- Get the line for this row
        local line = vim.api.nvim_buf_get_lines(bufnr, row, row + 1, false)[1]

        -- Check if it has trailing whitespace
        if #line == 0 or #line ~= #utils.get_whitespace(line) then
          return virt_text
        end

        -- Iterate over the virtual text
        for i, cell in ipairs(virt_text) do
          -- Iterate over each virtual text cell
          for j, highlight_group in ipairs(cell[2]) do
            -- Overwrite any highlight that is whitespace
            if vim.startswith(highlight_group, "@ibl.whitespace") then
              virt_text[i][2][j] = "Error" -- Change this to whatever highlight group you want
            end
          end
        end
        return virt_text
      end)
    end,
  },
  ```
  
Result is
![image](https://github.com/lukas-reineke/indent-blankline.nvim/assets/19948672/b8527825-6a10-490e-8a0b-86f8c7d8df24)
(same as with no hook)

**Reply by lukas-reineke** [reply]
No, that example will not work. The hook is only for trailing whitespace on empty lines that also have indentation.
Just trailing whitespace at the end of some text has nothing to do with indent blankline.
You can install another plugin to do that, like https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-trailspace.md for example.

**Comment by horseinthesky** [comment]
> No, that example will not work. The hook is only for trailing whitespace on empty lines that also have indentation. Just trailing whitespace at the end of some text has nothing to do with indent blankline. You can install another plugin to do that, like [echasnovski/mini.nvim@`main`/readmes/mini-trailspace.md](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-trailspace.md) for example.

Thank you for the explanation. I've found this small plugin that does the trick =)
https://github.com/johnfrankmorgan/whitespace.nvim
